ncm-ceph/src/main/perl/Ceph/Cluster.pm

```diff
@@ -0,0 +1,199 @@
+            $self->ssh_known_keys($host, $self->{key_accept}, $self->{cephusr});
+        }
+        if ($self->run_ceph_deploy_command([qw(gatherkeys), $host], 'gather ceph keyrings from monitors')) {
+            $ok = 1;
+            last;
+        }
+    }
+    if (!$ok) {
+        # Manual commands for new cluster  
+        # Run command with ceph-deploy for automation, 
+        # but take care of race conditions
+
+        my @newcmd = qw(new);
+        foreach my $host (@{$self->{init_hosts}}) {
+            push (@newcmd, $host);
+        }
+        if (!-f "$self->{cephusr}->{homeDir}/ceph.mon.keyring"){
+            $self->run_ceph_deploy_command([@newcmd], 'create new ceph cluster files');
+        }
+        $self->info("To create a new cluster, run this command");
+        $self->run_ceph_deploy_command([qw(mon create-initial)],'create initial monitors', printonly => 1, rwritecfg => 1);
+        return 0;
+    } else {
+        return 1;
+    }
+}
+
+# Fail if cluster not ready and no deploy hosts
+sub cluster_ready {
+    my ($self) = @_;
+
+    if (!$self->run_ceph_command([qw(status)], 'get cluster status' )) {
+            my @admin = ('admin', $self->{hostname});
+            $self->run_ceph_deploy_command(\@admin);
+            if (!$self->run_ceph_command([qw(status)], 'get cluster status')) {
+                # This should not happen
+                $self->error("Cannot connect to ceph cluster!");
+                return 0;
+            } else {
+                $self->debug(1,"Node ready to receive ceph-commands");
+            }
+    }
+    return 1;
+}
+
+sub write_init_cfg
+{
+    my ($self) = @_;
+    my $cfgfile = NCM::Component::Ceph::Cfgfile->new(
+        $self->{cfgtree}, $self, "$self->{prefix}/cluster/initcfg", $CEPH_DEPLOY_CFGFILE);
+    if (!$cfgfile->configure()) {
+         $self->error('Could not write cfgfile for ceph-deploy, aborting deployment');
+         return;
+    }
+    return 1;
+
+}
+
+sub prepare_cluster
+{
+    my ($self) = @_;
+
+    my $exists = $self->cluster_exists();
+    $self->write_init_cfg() or return;
+    return if !$exists;
+
+    $self->cluster_ready() or return;
+    
+    return 1;
+};
+
+sub test_host_connections
+{
+    my ($self, $map) = @_;
+    foreach my $host (keys %{$map->get_quattor_map}){
+        if (! $self->test_host_connection($map->get_fqdn($host), $self->{key_accept}, $self->{cephusr})){
+            $self->error("Can't reach ", $map->get_fqdn($host), " for configuration");
+            return;
+        }
+    }
+}
+
+# get running config and see what needs to be deployed
+sub make_tasks
+{
+    my ($self) = @_;
+    my $map = NCM::Component::Ceph::ClusterMap->new($self);
+    $map->map_existing() or return;
+    $map->map_quattor() or return;
+    $self->test_host_connections($map) or return;
+    return $map->get_deploy_map();
+}
+
+# Deploys a single daemon
+sub deploy_daemon 
+{
+    my ($self, $cmd, $name, $type) = @_;
+    push (@$cmd, $name);
```

Comment:
this modifies `$cmd`, is that what you really want?

Reply:
this is called with eg. deploy_daemon([qw(mon create)]

Reply:
it's still odd. use 
```perl
my ($self, $cmdref, $name, $type) = @_;
my @cmd = (@$cmdref, $name);
```